### PR TITLE
Backport code changes from GPU-related developments

### PR DESCRIPTION
**Pull request by fwyzard**
- add `const` accessors to conditions data
  - add `SOARotation` class and tests
  - add Phase 1 pixel topology test
  - BuildFile.xml clean up

### CODE REVIEW

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by fwyzard** [comment]
@perrotta, @ghellwig, @civanch, @Dr15Jones, @arunhep, @ianna, @mdhildreth, @franzoni, @cerminar, @slava77, @ggovi, @lpernie can you please review this, and comment ?

**Comment by Dr15Jones** [comment]
@ianna please take a look

File: Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#pragma once
+#ifndef Geometry_TrackerGeometryBuilder_phase1PixelTopology_h
+#define Geometry_TrackerGeometryBuilder_phase1PixelTopology_h
+
+#include <cstdint>
+
+namespace phase1PixelTopology {
+
+  constexpr uint16_t numRowsInRoc     = 80;
```
Comment (ianna) [inline Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: @fwyzard - shouldn't these hardcoded numbers be taken from an EventSetup nowadays?

Comment (VinInn) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: NO!
they are hardcoded for good reason: notice this is for phase1 and we have ONE and only ONE
phase1 pixel detector.

This is supposed to be used in very high performance computing. NO runtime ifs.
Only hardcoded constants.

Comment (ianna) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: @VinInn - the constants are iov dependant, the same as other conditions.

Comment (VinInn) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: constants are constants. please try to be realistic.
THESE numbers do not change, will never change. Are hardwired in the hardware down in the pit.
and in any case are Hardwired also in current code see for instance
https://cmssdt.cern.ch/dxr/CMSSW/source/Geometry/TrackerGeometryBuilder/src/RectangularPixelTopology.cc#85


The correct way of handling all this even with iov is to have all these numbers hardwired in their namespace as here, have High Performance algorithms templeted with that and choose using IOV and having the Low Performance analyzer (DQM) use a generic wrapper with ifs all over the place (as done in TrackerTopology)
TrackerTopology is not the way to go to get high performance

Comment (VinInn) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: to be clear, this file is just a HighPerformanceVersion of RectangularPixelTopology not much more hardwired only clearer actually while the "geometry constants" in RectangularPixelTopology are spread all over the place...

Comment (ianna) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: How will these constant relate to PTrackerParameters?

Comment (VinInn) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: Do you expect to modify PTrackerParameters?
In any case they relate to the real detector, better PTrackerParameters do the same.
So they relate to  PTrackerParameters as they originate from the very same source.
And again this file is supposed to emulate RectangularPixelTopology producing the very same results for Phase1 detector

Comment (ianna) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: IMHO, there should not be two ways to retrieve the same information. If this new high performant approach will be used based on IOV, perhaps, it should replace the PTrackerParameters.

Comment (VinInn) [reply Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h]: Sure, at some point. When we will have enough person power.
For the time being we are in R&D and we do not have the resources to fix legacy code.
CMS will need to decide, at some point, what to do will legacy code once we move to heterogeneous high-performance-coding for the upgrades.
Most probably we will need to maintain two (more?) sources of truth (and, as already pointed out with RectangularPixelTopology example, the current code is indeed full of multiple sources of truth)

**Comment by fwyzard** [comment]
@ianna thanks for your comments.

Indeed this raises a more general issue: do the experts here prefer that we integrate non-GPU related code in the CMSSW code-base and collect feedback on a regular basis (like this PR is trying to do) ?

The alternative is to keep all these changes in the Patatrack branch, and eventually come with a single PR in six months from now.

**Comment by slava77** [comment]
@fwyzard 
if it's a backport (although it looks more like a side-port) from somewhere, it would be nice to have a reference in the PR description.

File: DataFormats/GeometrySurface/interface/SOARotation.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
+#pragma once
+#ifndef DataFormats_GeometrySurface_SOARotation_h
```
Comment (slava77) [inline DataFormats/GeometrySurface/interface/SOARotation.h]: is ```pragma once``` really needed given that the include guard is present?

Comment (VinInn) [reply DataFormats/GeometrySurface/interface/SOARotation.h]: no. Actually I am trying to move to #pragma once and remove all those guards...

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/22409	42dc926
- reco changes are technical, restricted to RecoLocalTracker/SiPixelClusterizer/test/BuildFile.xml  which was just reformatted/reshuffled
- jenkins tests pass and comparisons with the baseline show no differences.


As for the general trickle integration, it makes sense if it's obvious that the changes are relatively final.

**Comment by fwyzard** [comment]
@slava77 thanks for the comments.

The ongoing development work is being integrated here: https://github.com/cms-patatrack/cmssw/ .
We are also maintaining instructions [here](https://patatrack.web.cern.ch/patatrack/wiki/patatrackdevelopment.html), and some performance comparisons [here](http://fwyzard.web.cern.ch/fwyzard/patatrack/20180303.095826/RelValZMM_13-CMSSW_10_1_0_pre1-100X_upgrade2018_realistic_v10-v1/plots_pixel.html), though we will try to have things more organised as we go along.

Indeed the changes that we are pushing to the upstream CMSSW are supposed to be final AFAIK; but I am also interested in getting comments from the relevant experts and L2's as early as possible.

File: DataFormats/GeometrySurface/interface/SOARotation.h (hunk 2 above)
Comment (slava77) [reply DataFormats/GeometrySurface/interface/SOARotation.h]: ``` #pragma once is a non-standard but widely supported preprocessor directive ```
https://en.wikipedia.org/wiki/Pragma_once

the current policy, AFAIK, is to use the include guards

Comment (davidlange6) [reply DataFormats/GeometrySurface/interface/SOARotation.h]: seems one wants #pragma once or include guards and not both.To start a migration towards #pragma might be best done with some discussion on the topic. [looks like its easy to find some interesting ones on google]

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by lpernie** [comment]
+1

**Comment by fwyzard** [comment]
Wait, this is *still pending* ?
It has been fully singed for *two weeks*, for an open development release, and still it was not merged ?
What is the purpose of asking five different groups of people to review and sign the changes, if then they are not integrated anyway ?

**Comment by VinInn** [comment]
Apparenly I can edit PR made by others...

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by fabiocos** [comment]
please test

**Pull request merged**